### ISSUE TITLE
export overload manager options at envoy

### ISSUE THREAD

**Issue by lambdai** [istio/istio#14366]
Among the overload actions this particular `envoy.overload_actions.shrink_heap` is helping proxy to reclaim the heap memory. It's extremely helpful when there is burst of config changes.

Ref:
https://www.envoyproxy.io/docs/envoy/latest/configuration/overload_manager/overload_manager#overload-actions

**Comment by howardjohn** [comment]
Nice, should be a pretty simple change to the bootstrap. Ill try it out

**Comment by ckcd** [comment]
@howardjohn 
Is this issue still valid ?  Can I take it ?

**Comment by howardjohn** [comment]
Sure, that would be great!

**Comment by lambdai** [comment]
Thank you for taking it @ckcd

**Comment by howardjohn** [comment]
If you want to apply this today:
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: istio-custom-bootstrap-config
data:
  custom_bootstrap.json: |
    {
      "stats_config": {
        "stats_matcher": {
          "inclusion_list": {
            "patterns": [
              {
              "prefix": "overload"
              },
            ]
          }
        }
      },
      "overload_manager": {
        "refresh_interval": {
          "seconds": 1,
        },
        "actions": [
          {
            "name": "envoy.overload_actions.shrink_heap",
            "triggers": [
            {
              "threshold": {
                "value": 0.90
              },
              "name": "envoy.resource_monitors.fixed_heap"
            }
            ]
          },
          {
            "name": "envoy.overload_actions.stop_accepting_requests",
            "triggers": [
            {
              "threshold": {
                "value": 0.99
              },
              "name": "envoy.resource_monitors.fixed_heap"
            }
            ]
          }
        ],
        "resource_monitors": [
          {
            "config": {
              "max_heap_size_bytes": 2147483648
            },
            "name": "envoy.resource_monitors.fixed_heap"
          }
        ]
      }
    }

```

Then follow other directions in `samples/custom-bootstrap/`

**Comment by zirain** [comment]
resource_monitors[0]) config: Cannot find field.) has unknown fields

please check you istio version and verify the configuration.

**Comment by style95** [comment]
I could apply the configuration with the following configs.

```yaml
$ cat custom-bootstrap-runtime.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: istio-custom-bootstrap-config
data:
  custom_bootstrap.json: |
    {
      "stats_config": {
        "stats_matcher": {
          "inclusion_list": {
            "patterns": [
              {
              "prefix": "overload"
              },
            ]
          }
        }
      },
      "overload_manager": {
        "refresh_interval": {
          "seconds": 1,
        },
        "actions": [
          {
            "name": "envoy.overload_actions.shrink_heap",
            "triggers": [
            {
              "threshold": {
                "value": 0.10
              },
              "name": "envoy.resource_monitors.fixed_heap"
            }
            ]
          },
          {
            "name": "envoy.overload_actions.stop_accepting_requests",
            "triggers": [
            {
              "threshold": {
                "value": 0.99
              },
              "name": "envoy.resource_monitors.fixed_heap"
            }
            ]
          }
        ],
        "resource_monitors": [
          {
            "typed_config": {
              "@type": "type.googleapis.com/envoy.extensions.resource_monitors.fixed_heap.v3.FixedHeapConfig",
              "max_heap_size_bytes": 2147483648
            },
            "name": "envoy.resource_monitors.fixed_heap"
          }
        ]
      }
    }
```

This part is updated.

```
"resource_monitors": [
          {
            "typed_config": {
              "@type": "type.googleapis.com/envoy.extensions.resource_monitors.fixed_heap.v3.FixedHeapConfig",
              "max_heap_size_bytes": 2147483648
            },
            "name": "envoy.resource_monitors.fixed_heap"
          }
```

**Comment by Pawan-Bishnoi** [comment]
This would be really helpful to have. I can give this a shot over the weekend (if we don't have this already)

**Comment by frittentheke** [comment]
> This would be really helpful to have. I can give this a shot over the weekend (if we don't have this already)

Before I open yet another issue @Pawan-Bishnoi  could you kindly update your view on this matter?
There was a PR (https://github.com/istio/istio/pull/17259) that never made it ... but since Envoy has such a great and feature rich overload manager (https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/overload/v3/overload.proto) and especially with edge gateways potentially being abused by outside traffic not having a good baseline overload config (see https://github.com/istio/istio/issues/24715 about adding best-practices) and even some knobs to tweak seems strange.
@howardjohn sorry to drag you into this comment to and old and already closed issue ... but we've been wondering if manually adding overload settings via custom bootstrap config and EnvoyFilters (see https://github.com/istio/istio/issues/24715#issuecomment-2511324032) is really the current state of things? Seems really like an opportunity missed for Istio to be better and more resilient for prime time than other ingress controllers / edge gateways.

**Comment by Pawan-Bishnoi** [comment]
Yes, I support adding first class support for this, ideally via ProxyConfig API.
Although that will only be useful if we dynamically set max_heap_size_bytes based on the actual pod/container spec.
That would allow us to configure overload protection mesh wide (with an option to override at proxy level).

**Comment by frittentheke** [comment]
I seem to be preaching to the choir at this point, but there is a recent and related incident at Spotify https://engineering.atspotify.com/2025/5/incident-report-spotify-outage-on-april-16-2025

Overload protection and heap size limits aligned with container limits should really avoid (most) traffic related OOM conditions.

**Comment by wujiaqi** [comment]
We recently added a new memory monitor to envoy which monitors memory usage in a cgroup aware way. We found some substantial flaws in the fixed heap monitor because the heap monitor reported by tcmalloc is not close enough to the memory that kubernetes sees so it is not very good for preventing OOMKills. I think this should be a supported mode at a minimum in the implementation given the kubernetes native nature of Istio. This Issue describes the problem and the associated PR is linked to it https://github.com/envoyproxy/envoy/issues/38718
